code/modules/food_and_drinks/recipes/soup_mixtures.dm

```diff
@@ -62,24 +62,45 @@
 	if(!length(required_ingredients))
 		return TRUE
 
-	// This is very unoptimized for something ran every handle-reaction for every soup recipe.
-	// Look into ways for improving this, cause bleh
+	//copy of all ingredients to check out
 	var/list/reqs_copy = required_ingredients.Copy()
+	//number of ingredients who's requested amounts has been satisfied
+	var/completed_ingredients = 0
 	for(var/obj/item/ingredient as anything in pot.added_ingredients)
-		// See if we fulfill all reqs
-		for(var/ingredient_type in required_ingredients)
-			if(!istype(ingredient, ingredient_type))
-				continue
-			if(isstack(ingredient))
-				var/obj/item/stack/stack_ingredient = ingredient
-				reqs_copy[ingredient_type] -= stack_ingredient.amount
+		var/ingredient_type = ingredient.type
+		do
+		{
+			var/ingredient_count = reqs_copy[ingredient_type]
+
+			//means we still have left over ingredients
+			if(ingredient_count)
+				//decode ingredient type i.e. stack or not and fulfill request
+				if(ispath(ingredient_type, /obj/item/stack))
+					var/obj/item/stack/stack_ingredient = ingredient
+					ingredient_count -= stack_ingredient.amount
+				else
+					ingredient_count -= 1
+
+				//assign final values
+				if(ingredient_count <= 0)
+					completed_ingredients += 1
+					ingredient_count = 0
+				reqs_copy[ingredient_type] = ingredient_count
+
+				//work complete
+				break
+
+			//means we have to look for subtypes
+			else if(isnull(ingredient_count))
+				ingredient_type = type2parent(ingredient_type)
+
+			//means we have no more remaining ingredients so bail, can happen if multiple ingredients of the same type/subtype are in the pot
 			else
-				reqs_copy[ingredient_type] -= 1
+				break
+		}
+		while(ingredient_type != /obj/item)
```

Comment:
@SyncIt21 I do not like this solution at all, the while loop stuff is too much, why not just use a (cached) typecache?

Reply:
https://github.com/tgstation/tgstation/pull/80021#discussion_r1409720108

It's ugly I know. But it's efficient. Also a typecache is a linear list and not a map of reagent type to it's required amount so it would require some hacks to make it work all for lesser performance

Reply:
I'm not suggesting type -> required amount, I'm suggesting type -> type looked for.

This while loop is gonna be expensive, and it leads to really like.. verbose? code. Generating lists for this should be trivial, and they're also only like 80 soup recipes so it really isn't a meaningful memory cost. equivalent to like maybe 100 extra turfs worth of gasmixtures.

Reply:
> I'm not suggesting type -> required amount, I'm suggesting type -> type looked for.

That's not how this code works. We require a map of type->required amount so that we can decrement those values as we match them with the pot ingredients and when it becomes 0 we can conclude we have all the ingredients and break. Have a look at the old code 

> This while loop is gonna be less efficient then that just by nature, and it leads to really like.. verbose?

Yes and also No. It depends on the soup ingredients. For example in the case of vegetable soups we require a carrot of type obj/item/food/carrot and botany can grow carrots of this exact type, so the while loop runs only once.
If we have requirments of long typepaths then Yes you are right the performance does degrade but compared to the performance of nested for loop it's still better.
In most of the cases(atleast as of now) we have a 1:1 match of ingredient type to required type and even if type paths do get longer the performance loss is still relatively better compared to before.

> Not worth caring about unless we start to get real memory issues, which we haven't had since we doubled our ram.

I'm not saying there isn't enough memory going around what I'm saying is if and when we get a chance to save memory we should take it. 
Also you are only looking at this only from the big server's perspective, there are people(like me) who hosts servers locally in our potato PC's and our specs aren't that great.

Reply:
> That's not how this code works. We require a map of type->required amount so that we can decrement those values as we match them with the pot ingredients and when it becomes 0 we can conclude we have all the ingredients and break. Have a look at the old code

The job of the while loop is to turn a typepath into its most recent parent that we require.
I am suggesting replacing this with a list lookup, into an assoc list that pairs all children with their most recent required parent.

This removes the need to iteratively reduce our type, and turns it into a cheap lookup.

> Yes and also No. It depends on the soup ingredients. For example in the case of vegetable soups we require a carrot of type obj/item/food/carrot and botany can grow carrots of this exact type, so the while loop runs only once. If we have requirments of long typepaths then Yes you are right the performance does degrade but compared to the performance of nested for loop it's still better. In most of the cases(atleast as of now) we have a 1:1 match of ingredient type to required type and even if type paths do get longer the performance loss is still relatively better compared to before.

Yeah it is better then the old system for sure, I just think it could be better is all. Especially if someone adds an ingredient that isn't required.

> I'm not saying there isn't enough memory going around what I'm saying is if and when we get a chance to save memory we should take it. Also you are only looking at this from the big server's perspective, there are people(like me) who hosts servers locally in our potato PC's and our specs aren't that great.

I don't think this is always the case. We trade memory for cpu all the time, and I don't think it's a meaningful impact here. You overestimate the weight of 80 lists.

I'm also not speaking in terms of like chonky server, when I say "we" doubled our ram I mean the limit at which byond will OOM doubled recently, from 2 gigs to 4.

Reply:
S not a huge deal this code just seems kinda off and i'd like to improve it if possible

Reply:
> if someone adds an ingredient that isn't required.

You got me there. Yes that would be bad. 

> You overestimate the weight of 80 lists. 

80 lists for now. In a few years as more recipes are added this statement will become a meme, I'm thinking ahead.

Overall I agree with everything your saying  but let's leave it here a case study if you will. 
It's soup not atmos code so ya know..life goes on